src/libasr/asr_verify.cpp

```diff
@@ -440,6 +440,25 @@ class VerifyVisitor : public BaseWalkVisitor<VerifyVisitor>
         verify_unique_dependencies(x.m_dependencies, x.n_dependencies,
                                    x.m_name, x.base.base.loc);
 
+        // Get the x symtab.
+        SymbolTable *sym = x.m_symtab->parent;
+
+        // Dependencies of the function should be from function's symbol table.
+        for( size_t i = 0; i < x.n_dependencies; i++ ) {
+            std::string found_dep = x.m_dependencies[i];
+
+            // Get the symbol of the found_dep.
+            ASR::symbol_t* dep_sym = sym->get_symbol(found_dep);
```

Comment:
I think this checks only upto the first parent of the current symbol table? May be if `sym->resolve_symbol(found_dep)` is `nullptr` then it means dependency is not imported as an `ExternalSymbol` so the check should fail? And if it is not `nullptr` then it means it belongs to one of the parents of `x.symtab` or is already imported as an `ExternalSymbol` (either of which is correct).

Reply:
In some test cases, the dependency symbol is nested inside the symbol table of x (if we look at asr of https://github.com/lcompilers/lpython/blob/main/integration_tests/test_numpy_03.py). In that case also, should the checks  fail ? As ``resolve_symbol`` will only look inside the parent and the current scope and not inside the child's scope so it will return ``nullptr`` for that too.

Reply:
I am confused. How can a symbol depend on something which is defined inside the symbol itself?

Reply:
We should not use `resolve_symbol`. I think the current code is correct.

What the verify() pass has to check is is that all dependencies are from the parent symbol table only. Never from any other symbol table.

Reply:
@czgdp1807 There is a bug while adding dependencies which needs to be fixed.

Reply:
Is there any example which I can use to reproduce the bug?

Reply:
You can try to run this example: 

```
from lpython import i32

def main():
    i: i32
    k: i32

    for k in range(10):
        i = k//3
        print(i)
```

Notice the ``_lpython_floordiv@__lpython_overloaded_6___lpython_floordiv`` is the dependency on which the main depends and it is nested inside the symbol table of ``main``.

Reply:
I will see this later.

Reply:
I will try to look at it by the end of this week once I fix some important fastGPT bugs in LFortran.

Reply:
For your example I get,

```zsh
(lp) 14:37:01:~/lpython_project/lpython % lpython /Users/czgdp1807/lpython_project/debug.py --show-asr --no-indent
(TranslationUnit (SymbolTable 1 {__main__: (Module (SymbolTable 2 {main: (Function (SymbolTable 3 {block: (Block (SymbolTable 4 {_lpython_floordiv: (ExternalSymbol 4 _lpython_floordiv 5 _lpython_floordiv lpython_builtin [] _lpython_floordiv Private), _lpython_floordiv@__lpython_overloaded_6___lpython_floordiv: (ExternalSymbol 4 _lpython_floordiv@__lpython_overloaded_6___lpython_floordiv 5 __lpython_overloaded_6___lpython_floordiv lpython_builtin [] __lpython_overloaded_6___lpython_floordiv Public)}) block [(= (Var 3 i) (FunctionCall 4 _lpython_floordiv@__lpython_overloaded_6___lpython_floordiv 4 _lpython_floordiv [((Var 3 k)) ((IntegerConstant 3 (Integer 4)))] (Integer 4) () ()) ()) (Print () [(Var 3 i)] () ())]), i: (Variable 3 i [] Local () () Default (Integer 4) () Source Public Required .false.), k: (Variable 3 k [] Local () () Default (Integer 4) () Source Public Required .false.)}) main (FunctionType [] () Source Implementation () .false. .false. .false. .false. .false. [] [] .false.) [] [] [(DoLoop () ((Var 3 k) (IntegerConstant 0 (Integer 4)) (IntegerBinOp (IntegerConstant 10 (Integer 4)) Sub (IntegerConstant 1 (Integer 4)) (Integer 4) (IntegerConstant 9 (Integer 4))) (IntegerConstant 1 (Integer 4))) [(BlockCall -1 3 block)])] () Public .false. .false. ())}) __main__ [lpython_builtin] .false. .false.), lpython_builtin: (IntrinsicModule lpython_builtin), main_program: (Program (SymbolTable 127 {}) main_program [] [])}) [])
(lp) 14:37:04:~/lpython_project/lpython % lpython /Users/czgdp1807/lpython_project/debug.py --show-asr --no-indent --pas
s=nested_vars
(TranslationUnit (SymbolTable 1 {__main__: (Module (SymbolTable 2 {main: (Function (SymbolTable 3 {block: (Block (SymbolTable 4 {_lpython_floordiv: (ExternalSymbol 4 _lpython_floordiv 5 _lpython_floordiv lpython_builtin [] _lpython_floordiv Private), _lpython_floordiv@__lpython_overloaded_6___lpython_floordiv: (ExternalSymbol 4 _lpython_floordiv@__lpython_overloaded_6___lpython_floordiv 5 __lpython_overloaded_6___lpython_floordiv lpython_builtin [] __lpython_overloaded_6___lpython_floordiv Public)}) block [(= (Var 3 i) (FunctionCall 4 _lpython_floordiv@__lpython_overloaded_6___lpython_floordiv 4 _lpython_floordiv [((Var 3 k)) ((IntegerConstant 3 (Integer 4)))] (Integer 4) () ()) ()) (Print () [(Var 3 i)] () ())]), i: (Variable 3 i [] Local () () Default (Integer 4) () Source Public Required .false.), k: (Variable 3 k [] Local () () Default (Integer 4) () Source Public Required .false.)}) main (FunctionType [] () Source Implementation () .false. .false. .false. .false. .false. [] [] .false.) [] [] [(DoLoop () ((Var 3 k) (IntegerConstant 0 (Integer 4)) (IntegerBinOp (IntegerConstant 10 (Integer 4)) Sub (IntegerConstant 1 (Integer 4)) (Integer 4) (IntegerConstant 9 (Integer 4))) (IntegerConstant 1 (Integer 4))) [(BlockCall -1 3 block)])] () Public .false. .false. ())}) __main__ [lpython_builtin] .false. .false.), lpython_builtin: (IntrinsicModule lpython_builtin), main_program: (Program (SymbolTable 127 {}) main_program [] [])}) [])
```

And both of them seem fine to me.

Reply:
Yes, should be now working fine in the current branch.